internal/controllers/machine/machine_controller.go

```diff
@@ -345,21 +345,30 @@ func (r *Reconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Clu
 
 			conditions.MarkTrue(m, clusterv1.DrainingSucceededCondition)
 			r.recorder.Eventf(m, corev1.EventTypeNormal, "SuccessfulDrainNode", "success draining Machine's node %q", m.Status.NodeRef.Name)
+		}
+
+		// After node draining is completed, and if isNodeVolumeDetachingAllowed returns True, make sure all
+		// volumes are detached before proceeding to delete the Node.
+		if r.isNodeVolumeDetachingAllowed(m) {
+			log.Info("Waiting for node volumes to be detached", "Node", klog.KRef("", m.Status.NodeRef.Name))
 
-			// After node draining, make sure volumes are detached before deleting the Node.
+			// The VolumeDetachSucceededCondition never exists before we wait for volume detachment for the first time,
+			// so its transition time can be used to record the first time we wait for volume detachment.
+			// This `if` condition prevents the transition time to be changed more than once.
 			if conditions.Get(m, clusterv1.VolumeDetachSucceededCondition) == nil {
 				conditions.MarkFalse(m, clusterv1.VolumeDetachSucceededCondition, clusterv1.WaitingForVolumeDetachReason, clusterv1.ConditionSeverityInfo, "Waiting for node volumes to be detached")
 			}
+
 			if ok, err := r.shouldWaitForNodeVolumes(ctx, cluster, m.Status.NodeRef.Name); ok || err != nil {
 				if err != nil {
-					r.recorder.Eventf(m, corev1.EventTypeWarning, "FailedWaitForVolumeDetach", "error wait for volume detach, node %q: %v", m.Status.NodeRef.Name, err)
+					r.recorder.Eventf(m, corev1.EventTypeWarning, "FailedWaitForNodeVolumesDetach", "error waiting for node volumes detaching, Machine's node %q: %v", m.Status.NodeRef.Name, err)
```

Comment:
```suggestion
					r.recorder.Eventf(m, corev1.EventTypeWarning, "FailedWaitForVolumeDetach", "error waiting for node volumes detaching, Machine's node %q: %v", m.Status.NodeRef.Name, err)
```

conditions error reasons are part of our API, and someone can build automation on top of it.
we should keep current reasons otherwise this is an API change that should wait for v1beta2

Reply:
Thanks, done!